### PR TITLE
[backend] Improve playbook automation

### LINKED ISSUE #6311: [Playbook] Error when applying label on some types of observables
## Description

The goal of this playbook is to apply a label on the observables of reports having that label.
The steps of the playbook are:

1. Listening to reports having a label "217"
image
2. Extracting the contained entities of the report
image
3. Filtering observables
image
4. Adding the label "217"
image
5. Send to ingestion

## Environment

Testing

## Reproducible Steps

Steps to create the smallest reproducible scenario:
1. Create a new empty report with the label "217"
2. Add an Artefact (make sure you are adding one that doesn't have previously the label "217")
3. It works!
4. Add a URL (make sure you are adding one that doesn't have previously the label "217")
5. Error in the playbook

## Expected Output

The label is added to the URL

## Actual Output

No label added + error at the level of "Manipulate knowledge":
`{
  "message": "Cannot set properties of undefined (setting 'labels')",
  "stack": "TypeError: Cannot set properties of undefined (setting 'labels')\n    at Object.add (/opt/opencti/build/node_modules/fast-json-patch/commonjs/core.js:14:16)\n    at applyOperation (/opt/opencti/build/node_modules/fast-json-patch/commonjs/core.js:235:60)\n    at Object.applyPatch (/opt/opencti/build/node_modules/fast-json-patch/commonjs/core.js:280:22)\n    at Object.executor (/opt/opencti/build/src/modules/playbook/playbook-components.ts:573:17)\n    at playbookExecutor (/opt/opencti/build/src/manager/playbookManager.ts:135:44)\n    at playbookExecutor (/opt/opencti/build/src/manager/playbookManager.ts:202:15)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at playbookExecutor (/opt/opencti/build/src/manager/playbookManager.ts:202:9)\n    at playbookExecutor (/opt/opencti/build/src/manager/playbookManager.ts:202:9)\n    at playbookStreamHandler (/opt/opencti/build/src/manager/playbookManager.ts:268:13)\n    at processStreamResult (/opt/opencti/build/src/database/redis.ts:600:3)\n    at processStep (/opt/opencti/build/src/database/redis.ts:654:31)\n    at processingLoop (/opt/opencti/build/src/database/redis.ts:671:13)",
  "name": "TypeError"
}`

## Additional information

There is another bug that I showed in the end of video below: When an error happens in a playbook, the whole content (steps, logs and icons) of the previous executions is replaced by the last one. This is why you can see the 4th step of the previous execution change into red.
Seen with @nino-filigran. I can create a new issue for this bug.

## Screenshots (optional)
playbook_inherit_label

### LINKED ISSUE #6739: [Playbook] Rename Filter knowledge to Match to reflect his real behavior
## Description

I am trying to create a playbook to automatically add a label to the CVE contained in a report.
While using a Filter knowledge, the following action will be executed on all the entities no matter the filter
Workaround with Reduce knowledge works fine

reviewed with 

## Environment

cloud demo environment

## Reproducible Steps

Steps to create the smallest reproducible scenario:
1. create a new event
2. create a listener
image
3. apply predefined rule
image
4. create filter knowledge
image
5. Manipulate knowledge
image
6. send for ingestion

Overall playbook:
image

## Expected Output

label applied only to the vulnerabilities

## Actual Output
While creating/editing a report, the label is applied to all the entities (not only the vulnerabilities):
image

### LINKED ISSUE #6796: [Playbook] Detection field not updated by the playbook
## Description

When I try to modify the "detection" field of my Indicators with playbooks, it's not working. The execution logs are correct but the update is not done.

## Environment

OCTI 6.0.10

## Reproducible Steps

Steps to create the smallest reproducible scenario:
1. Create a playbook with
- Listener : filter on "Entity type: Indicator AND label: test" (create and update event)
- Manipulate : replace - detection - true
- send to ingestion
 
2. Go on an Indicator and apply the label "test"
3. Check the execution log -> everything seem ok
4. Check the indicator -> detection field still false

## Screenshots (optional)

Screenshot 2024-04-23 222114

### LINKED ISSUE #6866: [Playbook] Generation of indicator after enrichment does not work
## Description

In my playbook, when I try to generate indicators from observables, i.e. "Promote" component, after enrichment, an error occurs.

The problem seems to come only if I generate them after enrichment, because if I reverse the order of execution (I put the component for generating indicators before the one for enrichment), I don't have the problem.

## Environment

OCTI 6.0.10

## Reproducible Steps

Steps to create the smallest reproducible scenario:
1. Create a playbook with
- Listener on IPv4 (observable) and label:test
- Enrich with virustotal
- Promote observable (switch on the "Create indicators from all observables in the bundle" toggle)
- Send to ingestion
2. Apply the label "test" on an IPv4
3. Look at the execution logs of the playbook -> an error has occurred on the "Promote" component

## Additional information

If you switch the component 2 and 3 (i.e. "promote" before "enrich", it works (no error).

### Playbook which doesn't work
Screenshot 2024-05-02 150825

### Listener config
Screenshot 2024-05-02 150903

### Promote observable config
Screenshot 2024-05-02 150851

### The error on the "Promote" component
Screenshot 2024-05-02 150739

### LINKED ISSUE #6941: [Playbook] Manipulate knowledge/manipulate all elements in the bundle Error
## Description

When executing a playbook, if there is an observable and an indicator in the bundle, adding a tag will result in this error.

image

## Environment

1. OS (where OpenCTI server runs): { e.g. Mac OS 10, Windows 10, Ubuntu 16.4, etc. }
2. OpenCTI version: 6.1.0
3. Other environment details: also tested on testing environment

## Reproducible Steps

Steps to create the smallest reproducible scenario:
1. Create playbook: https://demo.octi.filigran.io/dashboard/data/processing/automation/f1e38e99-ed1d-4562-8dfa-787212ecd0b8
image
Listen to observable creation, promote to indicator, add label, ingest
2. execution shows error

If "Manipulate all elements in the bundle" is disabled, there is no error but label is applied to the observable only
image

## Expected Output

Label applied to both observable and indicator

## Actual Output

{
  "message": "Type undefined not supported",
  "stack": "DATABASE_ERROR: Type undefined not supported\n    at error (/opt/opencti/build/src/config/errors.js:8:10)\n    at DatabaseError (/opt/opencti/build/src/config/errors.js:58:48)\n    at getParentTypes (/opt/opencti/build/src/schema/schemaUtils.js:151:11)\n    at /opt/opencti/build/src/modules/playbook/playbook-components.ts:567:80\n    at Array.filter (<anonymous>)\n    at computeAttributePath (/opt/opencti/build/src/modules/playbook/playbook-components.ts:567:66)\n    at /opt/opencti/build/src/modules/playbook/playbook-components.ts:581:30\n    at Array.map (<anonymous>)\n    at Object.executor (/opt/opencti/build/src/modules/playbook/playbook-components.ts:580:12)\n    at playbookExecutor (/opt/opencti/build/src/manager/playbookManager.ts:137:44)\n    at playbookExecutor (/opt/opencti/build/src/manager/playbookManager.ts:205:15)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at playbookExecutor (/opt/opencti/build/src/manager/playbookManager.ts:205:9)\n    at playbookStreamHandler (/opt/opencti/build/src/manager/playbookManager.ts:279:17)\n    at processStreamResult (/opt/opencti/build/src/database/redis.ts:642:3)\n    at processStep (/opt/opencti/build/src/database/redis.ts:696:31)\n    at processingLoop (/opt/opencti/build/src/database/redis.ts:713:13)",
  "name": "DATABASE_ERROR"
}

### LINKED ISSUE #6978: [Playbook] Issue when manipulating an observable
## Description

When creating a playbook that manipulates observables with no score an error is displayed. (perhaps similar to https://github.com/OpenCTI-Platform/opencti/issues/6941)

When manually adding a score or modifying the description the step goes well.

## Environment

1. OpenCTI 6.1.0

## Reproducible Steps

Steps to create the smallest reproducible scenario:
1. Create and start a playbook listening IPV4,
2. Create an IPV4 with no score,
3. The playbook starts but stop early at stage 'manipulate knowledge' displaying an error.

## Expected Output

Knowledge manipulation, such as adding a label or marking definition, should run smoothly.
## Actual Output

If no score is defined, the "manipulate knowledge" step will not work.
## Screenshots (optional)
image

### PR DESCRIPTION
**Pull request by richard-julien**
Improve playbook automation 
#6978
#6941
#6796 
#6739
#6866
#6311

### CODE REVIEW

File: opencti-platform/opencti-graphql/src/modules/playbook/playbook-components.ts [hunk 1]
```diff
@@ -543,6 +545,8 @@ const PLAYBOOK_UPDATE_KNOWLEDGE_COMPONENT: PlaybookComponent<UpdateConfiguration
   configuration_schema: PLAYBOOK_UPDATE_KNOWLEDGE_COMPONENT_SCHEMA,
   schema: async () => PLAYBOOK_UPDATE_KNOWLEDGE_COMPONENT_SCHEMA,
   executor: async ({ dataInstanceId, playbookNode, bundle }) => {
+    const context = executionContext('playbook_components');
+    const cacheIds = await getEntitiesMapFromCache(context, AUTOMATION_MANAGER_USER, ENTITY_TYPE_MARKING_DEFINITION);
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-graphql/src/modules/playbook/playbook-components.ts:573]: what is `cacheIds` ? the map of marking definitions from cache ? if it's only marking definitions, why do we call it `cacheIds` ?

File: opencti-platform/opencti-graphql/src/modules/playbook/playbook-components.ts [hunk 2]
```diff
@@ -553,10 +557,9 @@ const PLAYBOOK_UPDATE_KNOWLEDGE_COMPONENT: PlaybookComponent<UpdateConfiguration
       return undefined;
     };
     // Compute if attribute is defined as numeric
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-graphql/src/modules/playbook/playbook-components.ts]: maybe update the comment ? or remove it ?

File: opencti-platform/opencti-graphql/src/modules/playbook/playbook-components.ts [hunk 3]
```diff
@@ -580,26 +588,42 @@ const PLAYBOOK_UPDATE_KNOWLEDGE_COMPONENT: PlaybookComponent<UpdateConfiguration
...
-            path,
-            // eslint-disable-next-line no-nested-ternary,max-len
-            value: multiple ? action.value.map((o) => (numeric ? Number(o.patch_value) : o.patch_value)) : numeric ? Number(R.head(action.value)?.patch_value) : R.head(action.value)?.patch_value
-          }));
+          .map(({ action, path, multiple, attributeType }) => {
+            if (multiple) {
+              const currentValues = jsonpatch.getValueByPointer(bundle, path) ?? [];
+              const actionValues = action.value.map((o) => {
+                // If value is an id, must be converted to standard_id has we work on stix bundle
+                if (cacheIds.has(o.patch_value)) return (cacheIds.get(o.patch_value) as BasicStoreCommon).standard_id;
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-graphql/src/modules/playbook/playbook-components.ts:625]: what is cacheIds ? I don't understand this part.

Comment (richard-julien) [reply opencti-platform/opencti-graphql/src/modules/playbook/playbook-components.ts:625]: Related to // If value is an id, must be converted to standard_id has we work on stix bundle
We have filter like markings using internal_id for configuration. They need to be translated to standard to allow correct filtering

**Pull request merged**